### PR TITLE
RavenDB-19117

### PR DESCRIPTION
**Pull request by ayende**
- Corax - optimize term insertion using locality
- Avoid pinvoke call that we can easily inline
- Suppress security check for Zstd native calls

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19117

### Additional description

We currently pay a lot in terms of performance during indexing due to lookup of terms. We can probably reduce that cost by utilizing term locality to handle that.

### Type of change

- Optimization

### How risky is the change?

- Moderate 

Major change, but in experimental code only. 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### CODE REVIEW

**Comment by arekpalinski** [comment]
@ayende After merging other PRs we have conflicts here. Can you please have a look?

**Comment by ayende** [comment]
Yes, should be fixed now :-)
Also with (hopefully) fully working on the TB range dataset

File: src/Sparrow/Utils/ZstdLib.cs [hunk 1]
```diff
@@ -2,11 +2,13 @@
 using System.Diagnostics;
 using System.IO;
 using System.Runtime.InteropServices;
+using System.Security;
 using Sparrow.Platform;
 
 namespace Sparrow.Utils
 {
 #if NETCOREAPP3_1_OR_GREATER
+    [SuppressUnmanagedCodeSecurity]
```
Comment (arekpalinski) [inline src/Sparrow/Utils/ZstdLib.cs:11]: How much do we benefit from this?

Comment (arekpalinski) [reply src/Sparrow/Utils/ZstdLib.cs:11]: For reference: https://docs.microsoft.com/en-us/dotnet/api/system.security.suppressunmanagedcodesecurityattribute?view=net-6.0

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 2]
```diff
@@ -426,10 +426,113 @@ public bool TryGetValue(ReadOnlySpan<byte> key, out long value, out EncodedKey e
...
+                encodedKey = EncodedKey.Get(key, this, state.Header->DictionaryId);
+                encodedKey = FindPageFor(ref _internalCursor, ref state, encodedKey);
+                state = ref _internalCursor._stk[_internalCursor._pos];
+
+                if (state.LastMatch == 0) // found it
+                    return ReturnValue(encodedKey, ref state, out value);
+                // did *not* find it, but we are somewhere on the tree that is ensured
+                // to be at the key location *or before it*, so we can now start scanning *up*
+            }
+            Debug.Assert(state.Header->PageFlags == CompactPageFlags.Leaf);
```
Comment (arekpalinski) [inline src/Voron/Data/CompactTrees/CompactTree.cs]: Let's add a message argument to `Debug.Assert()` and put actual `state.Header->PageFlags` value there so we'll see it when it fails.

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 3]
```diff
@@ -426,10 +426,113 @@ public bool TryGetValue(ReadOnlySpan<byte> key, out long value, out EncodedKey e
...
+            var shouldBeInCurrentPage = pos < state.Header->NumberOfEntries;
+
+            if (shouldBeInCurrentPage)
+            {
+                var nextEntry = GetEncodedKey(state.Page, state.EntriesOffsets[pos]);
+                var match = encodedKey.Encoded.SequenceCompareTo(nextEntry);
+                shouldBeInCurrentPage = match < 0;
+            }
+
+            if (shouldBeInCurrentPage == false)
```
Comment (arekpalinski) [inline src/Voron/Data/CompactTrees/CompactTree.cs:494]: Minor - wouldn't be simpler to have `else` here instead?

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 4]
```diff
@@ -426,10 +426,113 @@ public bool TryGetValue(ReadOnlySpan<byte> key, out long value, out EncodedKey e
...
+            encodedKey = EncodedKey.Get(key, this, state.Header->DictionaryId);
+
+            SearchInCurrentPage(encodedKey, ref state);
+            if (state.LastSearchPosition  >= 0) // found it, yeah!
+            {
+                value = GetValue(ref state, state.LastSearchPosition);
+                return true;
+            }
+
+            var pos = ~state.LastSearchPosition;
```
Comment (arekpalinski) [inline src/Voron/Data/CompactTrees/CompactTree.cs:484]: Just to make sure - is it intentional to use `~` operator here? If yes then it seems to be tricky - can you elaborate on that please?

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 5]
```diff
@@ -741,7 +847,9 @@ public void Add(ReadOnlySpan<byte> key, long value)
         public void Add(ReadOnlySpan<byte> key, long value, EncodedKey encodedKey)
         {
             AssertValueAndKeySize(key, value);
-
+            // this overload assumes that a previous call to TryGetValue (where you go the encodedKey
+            // already placed us in the right place for the value)
+            Debug.Assert(_internalCursor._stk[_internalCursor._pos].Header->PageFlags == CompactPageFlags.Leaf);
```
Comment (arekpalinski) [inline src/Voron/Data/CompactTrees/CompactTree.cs]: Let's add the value of `_internalCursor._stk[_internalCursor._pos].Header->PageFlags` to the assertion message

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Voron/Data/CompactTrees/CompactTree.cs:494]: No, I see that `shouldBeInCurrentPage` is modified inside above if: `shouldBeInCurrentPage = match < 0;`

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 4 above)
Comment (arekpalinski) [reply src/Voron/Data/CompactTrees/CompactTree.cs:484]: I see we use it the same in other places

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: done

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 5 above)
Comment (arekpalinski) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: done

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 4 above)
Comment (ayende) [reply src/Voron/Data/CompactTrees/CompactTree.cs:484]: Yes, that is a common approach in BinarySearch lookup, we return negative number that is the `~` of the _expected_ position

File: src/Sparrow/Utils/ZstdLib.cs (hunk 1 above)
Comment (ayende) [reply src/Sparrow/Utils/ZstdLib.cs:11]: We hope a lot, need to profile again on the big machine

**Pull request merged**